omnigibson/scene_graphs/graph_exporter.py

```diff
@@ -0,0 +1,90 @@
+import json
+
+import numpy as np
+
+from igibson.scene_graphs.enum_obj_states import BinaryStatesEnum, UnaryStatesEnum
+from igibson.scene_graphs.graph_builder import SceneGraphBuilder
+
+
+class SceneGraphExporter(SceneGraphBuilder):
+    def __init__(self, h5py_file, num_frames_to_save=None, full_obs=False, **kwargs):
+        """
+        @param h5py_file: file to store exported scene graph
+        @param num_frames_to_save: total number of equidistant frames to save. None if you want to set all frames.
+        """
+        if num_frames_to_save is None:
+            assert full_obs, "You can only use full observability mode if you're subsampling frames."
+
+        super(SceneGraphExporter, self).__init__(only_true=True, full_obs=full_obs, **kwargs)
+        self.h5py_file = h5py_file
+
+        self.num_frames_to_save = num_frames_to_save
+        self.frame_idxes_to_save = None
+
+    def start(self, activity, log_reader):
+        super(SceneGraphExporter, self).start(activity, log_reader)
+
+        scene = activity.simulator.scene
+        objs = set(scene.objects()) | set(activity.object_scope.values())
+        self.num_obj = len(objs)
+        self.dim = 34  # 34: dimension of all relevant information
+        # create a dictionary that maps objs to ids
+        self.obj_to_id = {obj: i for i, obj in enumerate(objs)}
+
+        self.h5py_file.attrs["id_to_category"] = json.dumps(
+            {self.obj_to_id[obj]: obj.category for obj in self.obj_to_id}
+        )
+        self.h5py_file.attrs["id_to_name"] = json.dumps({self.obj_to_id[obj]: obj.name for obj in self.obj_to_id})
+
+        if self.num_frames_to_save is not None:
+            key_presses = log_reader.hf["agent_actions"]["vr_robot"][:, [19, 27]]
+            assert len(key_presses) > 0, "No key press found: too few frames."
+            any_key_press = np.max(key_presses[200:], axis=1)
+            first_frame = np.argmax(any_key_press) + 200
+            assert np.any(key_presses[first_frame] == 1), "No key press found: robot never activated."
+            self.frame_idxes_to_save = set(
+                np.linspace(
+                    first_frame, log_reader.total_frame_num - 1, self.num_frames_to_save, endpoint=True, dtype=int
+                )
+            )
+
+    def step(self, activity, log_reader):
+        frame_count = activity.simulator.frame_count
+        if self.num_frames_to_save is not None and frame_count not in self.frame_idxes_to_save:
+            return
+
+        super(SceneGraphExporter, self).step(activity, log_reader)
```

Comment:
nit -- in general, shouldn't we use python 3+ syntax? it's a bit more robust i think, and handles multiple inheritance implicitly. e.g.: `super().method(...)`

Reply:
Yes sorry this was never run on Python 3 (I just copied it over from iG) - fixed.